src/lucky_record/repo.cr

```diff
@@ -1,23 +1,57 @@
 class LuckyRecord::Repo
+  alias FiberId = UInt64
+
   @@db : DB::Database? = nil
+  private class_getter transactions = {} of FiberId => DB::Transaction
 
   Habitat.create do
     setting url : String
     setting lazy_load_enabled : Bool = true
   end
 
   def self.run
-    yield db
+    yield current_transaction.try(&.connection) || db
   end
 
   def self.db
     @@db ||= DB.open(settings.url)
   end
 
+  def self.current_transaction : DB::Transaction?
+    transactions[Fiber.current.object_id]?
+  end
+
   def self.truncate
     DatabaseCleaner.new.truncate
   end
 
+  # Rollback the current transaction
+  def self.rollback
+    raise LuckyRecord::Rollback.new
+  end
+
+  def self.transaction : Bool
+    if current_transaction
+      yield
+    else
+      wrap_in_transaction do
+        yield
+      end
+    end
+  end
+
+  def self.wrap_in_transaction
```

Comment:
@mikeeus I saved the current transaction to a map with the current fiber id as the key. That gets it working with nested transactions, and avoids passing the transaction all the time.

Let me know if you have any questions about the implementation!

Reply:
This is super cool! I knew the implementation would need to use `Fibers` but I wasn't sure how.